validator_client/src/config.rs

```diff
@@ -1,32 +1,131 @@
+        let data_dir: PathBuf = match arguments.value_of("datadir") {
+            Some(path) => PathBuf::from(path.to_string()),
+            None => {
+                let home = dirs::home_dir().ok_or_else(|| {
+                    Error::new(ErrorKind::NotFound, "Unable to determine home directory.")
+                })?;
+                home.join(DEFAULT_VALIDATOR_DATADIR)
+            }
         };
-        fs::create_dir_all(&data_dir)
-            .unwrap_or_else(|_| panic!("Unable to create {:?}", &data_dir));
-        let server = "localhost:50051".to_string();
-        let spec = ChainSpec::foundation();
-        Self {
+        fs::create_dir_all(&data_dir)?;
+
+        let validator_dir = data_dir.join(DEFAULT_VALIDATORS_SUBDIR);
+        fs::create_dir_all(&validator_dir)?;
+
+        let server: String = match arguments.value_of("server") {
+            Some(srv) => {
+                //TODO: I don't think this parses correctly a server & port combo
+                srv.parse::<u16>()
+                    .map_err(|e| Error::new(ErrorKind::InvalidInput, e))?
+                    .to_string()
+            }
+            None => "localhost:50051".to_string(),
+        };
+
+        // TODO: Permit loading a custom spec from file.
+        let spec: ChainSpec = match arguments.value_of("spec") {
+            Some(spec_str) => {
+                match spec_str {
+                    "foundation" => ChainSpec::foundation(),
+                    "few_validators" => ChainSpec::few_validators(),
+                    // Should be impossible due to clap's `possible_values(..)` function.
+                    _ => unreachable!(),
+                }
+            }
+            None => ChainSpec::foundation(),
+        };
+
+        Ok(Self {
             data_dir,
+            validator_dir,
             server,
             spec,
+        })
+    }
+
+    /// Try to load keys from validator_dir, returning None if none are found or an error.
+    pub fn fetch_keys(&self) -> Result<Option<Vec<Keypair>>, Error> {
+        let mut validator_dirs = fs::read_dir(&self.validator_dir)?.peekable();
+
+        // There are no validator directories.
+        if validator_dirs.peek().is_none() {
+            return Ok(None);
+        }
+
+        let mut key_pairs: Vec<Keypair> = Vec::new();
+
+        for validator_dir_result in validator_dirs {
+            let validator_dir = validator_dir_result?;
+
+            // Try to open the key file directly
+            // TODO skip keyfiles that are not found, and log the error instead of returning it.
+            let mut key_file = File::open(validator_dir.path().join(DEFAULT_PRIVATE_KEY_FILENAME))?;
+
+            let key: Keypair = bincode::deserialize_from(&mut key_file)
+                .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
+
+            // TODO skip keyfile if it's not matched, and log the error instead of returning it.
+            let validator_directory_name =
```

Comment:
If one directory is bad.. it will not just log the error. It will terminate and stop all future directories from being processed. Perhaps you may just want to skip this directory instead, ignore or propagate up afterwards and continue processing others?